Adding support for periodic Gmsh meshes [gmsh-periodic-dev]

mlstowell opened the pull request:
This PR supports reading the `$Periodic` section of Gmsh MSH ASCII files.  Both versions 2.2 and 4.1 have been tested.  The mesh reader applies the master/slave vertex mapping defined in the MSH file and creates a discontinuous, first order L2 field to represent the mesh coordinates in the usual MFEM manner.

<!--GHEX{"id":1478,"author":"mlstowell","editor":"tzanio","reviewers":["jandrej","bslazarov"],"assignment":"2020-05-14T17:56:13-07:00","approval":"2020-06-01T02:56:13.719Z","merge":"2020-06-07T19:22:53.509Z"}XEHG--><!--GHEXTABLE-->
 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge| 
 | --- | --- | --- | --- | --- | --- | --- | 
| [#1478](https://github.com/mfem/mfem/pull/1478) | @mlstowell | @tzanio | @jandrej + @bslazarov | 05/14/20 | 05/31/20 | 06/07/20 | |
<!--ELBATXEHG-->

tzanio commented:
@jandrej and @bslazarov, reminder that the review for this PR is due soon

jandrej commented on mesh/mesh_readers.cpp:
```diff
@@ -1291,6 +1291,64 @@ void Mesh::ReadGmshMesh(std::istream &input)
+         // Convert nodes to discontinuous GridFunction
+         this->SetCurvature(1, true, Dim, Ordering::byVDIM);
+
+         // Renumber elements to remove slave vertices
```
Could you elaborate what this does? Lets say we are looking at a rectangle (0,1)x(0,1) which is periodic at x=0 and x=1.

bslazarov replied:
I am having troubles running ex1 with mesh generated by the following geo file:

SetFactory("OpenCASCADE");
  
R = 2;
Torus(2) = {3,0,0, 1.5, 1, Pi/3};

pts() = PointsOf{ Volume{2}; };

Characteristic Length{ pts() } = 0.4;

// Characteristic Length{ 2 } = 0.01;

// Find bottom and top surfaces using a bounding box search:
e = 1e-6;
bottom() = Surface In BoundingBox{-e,-e,-e, R+e,R+e,e};
top() = Surface In BoundingBox{-e,-e,R-e, R+e,R+e,R+e};

// Set a rotation periodicity constraint:
Periodic Surface{9} = {8} Rotate{{0,0,1}, {3,0,0}, Pi/3 };

I think a test example in necessary.

mlstowell replied:
@jandrej , are you suggesting I expand the comment for this line to explain MFEM's strategy for handling periodic meshes?  I can certainly do that if you like.

mlstowell replied:
@bslazarov , an example is a good idea.  I can add something simple.  I tried your geo file in Gmsh 4.5.6 and got an error regarding surfaces 9 and 8 not being found.  I don't work with Gmsh myself.  The examples I was working from were provided to me by one of the researchers in FES.

jandrej replied:
I was just curious about the "remove slave vertices" comment.

mlstowell replied:
As I look at it again I see that comment is not clear.  I'll reword it.  What happens in the following loops is that the element and boundary element connectivity are altered to replace all references to the "slave" vertices with references to the corresponding "master" vertices.  This effectively stitches the two periodic surfaces together topologically.

bslazarov replied:
I have messed up the last line :-)

Here is the correct one:
// Set a rotation periodicity constraint:
Periodic Surface{3} = {2} Rotate{{0,0,1}, {3,0,0}, Pi/3 };

mlstowell replied:
Hi, @bslazarov , the trouble was not with the periodicity but with the element and boundary attributes which are equal to zero by default.  Adding the `Physical [Surface|Volume] ...` commands towards the end of the geo file seems to help.

```
SetFactory("OpenCASCADE");

R = 2;
Torus(1) = {3,0,0, 1.5, 1, Pi};

pts() = PointsOf{ Volume{1}; };

Characteristic Length{ pts() } = 0.4;

// Set a rotation periodicity constraint:
Periodic Surface{3} = {2} Rotate{{0,0,1}, {3,0,0}, Pi };

// Tag surfaces and volumes with positive integers
Physical Surface(1) = {1};
Physical Surface(2) = {2};
Physical Surface(3) = {3};
Physical Volume(1) = {1};

// Generate 3D mesh
Mesh 3;
```

mlstowell replied:
Actually there was a periodicity issue as well.  If the characteristic length is too large there will be fewer than three edges connecting the periodic surfaces and MFEM cannot always distinguish such edges.

bslazarov replied:
Thanks

bslazarov replied:
@mlstowell  - it seems that I am getting the same errors again, even if I tag the boundaries. I am getting the following errors:

MFEM abort: Gmsh file : vertices indices are not unique
 ... in function: void mfem::Mesh::ReadGmshMesh(std::istream&, int&, int&)
 ... in file: /home/blaz/develop/mark/mfem/mesh/mesh_readers.cpp:952

 The error appear for non-periodic mesh also. I am using 4.1 format for gmsh - let me know if that is the issue.

mlstowell replied:
@jandrej , working with @bslazarov I just added 2D and 3D example meshes to the `data` subdirectory and new sample runs to exercise the new code.

mlstowell commented:
@bslazarov found a memory leak in the Gmsh mesh reader which was related to `MFEM_USE_MEMALLOC` and the use of the `TetMemory` data structure.  I found a similar bug in the Cubit mesh reader and decided to fix it in this PR because it is such a small change.

tzanio commented on data/annulus-pi-3.msh:
```diff
@@ -0,0 +1,236 @@
+$MeshFormat
```
This will be the first time we add Gmsh files in the repository. Do we want to do that, or is https://github.com/mfem/data/tree/master/gmsh a better place for them?

tzanio commented:
Merged in `next` for testing...

tzanio commented:
Error from tonight's autotest run on toss:

```
mesh/mesh_readers.cpp(1954): error: expected a ";"
                 elements[elcount] = TetMemory.Alloc());
                                                      ^

compilation aborted for mesh/mesh_readers.cpp (code 2)
make[3]: *** [mesh/mesh_readers.o] Error 2
make[3]: *** Waiting for unfinished jobs....
make[2]: *** [/usr/WS2/mfem/autotest/toss_3_x86_64_ib/mfem/libmfem.so] Error 2
 **** MFEM build failed.
make[1]: *** [logs-toss_3_x86_64_ib/mfem-build.log] Error 2
make: *** [mfem-full] Error 2
mesh/mesh_readers.cpp(1954): error: expected a ";"
                 elements[elcount] = TetMemory.Alloc());
                                                      ^

compilation aborted for mesh/mesh_readers.cpp (code 2)
make: *** [mesh/mesh_readers.o] Error 2
make: *** Waiting for unfinished jobs....
mesh/mesh_readers.cpp(1954): error: expected a ";"
                 elements[elcount] = TetMemory.Alloc());
                                                      ^

compilation aborted for mesh/mesh_readers.cpp (code 2)
make[1]: *** [mesh/mesh_readers.o] Error 2
make: *** [test] Error 2
```

tzanio commented:
Re-merged in `next` for testing ...

mlstowell commented:
I hadn't noticed how large the Gmsh torus mesh was so I replaced it with a smaller version.  I also followed @bslazarov 's advice and included the geo files.  Finally, I added sample runs to `ex11p` since these showcase the periodic structure of the mesh better than `ex1p` can.

v-dobrev commented:
I saw a small visualization issue: when I tried `./ex1 -m ../data/torus-pi-3.msh` the periodic circular boundaries are not displayed. However, if I just open the mesh in GLVis those boundaries are displayed. Any idea why this is happening?

tzanio commented:
Re-merged in `next` for testing ...

mlstowell commented:
@v-dobrev , no, I'm not really sure.  The torus mesh does not define boundary elements on those circular surfaces so I assume this is why GLVis doesn't recognize these surfaces as needing to be painted.  Of course we cannot add these boundary elements to the mesh or `ex1p` will impose Dirichlet BCs on them.  I presume that mesh visualization in GLVis is not based on boundary elements.

v-dobrev commented:
@mlstowell, I found out what's going on: the reason for the difference is that I was using GLVis built with MFEM/master -- the Gmsh reader there was importing the mesh differently. With GLVis built with this branch, the periodic boundary is not seen when viewing the mesh file directly in GLVis.

We can add the periodic boundaries in the files (at least for visualization purposes) but you are right that `ex1`/`ex1p` will impose Dirichlet b.c. there too. We can probably add a method in MFEM that un-tags internal boundaries (periodic boundaries are seen as internal) and use it in `ex1` but we can do that another time.

mlstowell commented:
I'm not sure about curved periodic meshes.  I don't think I've seen Gmsh's curved mesh format.  I imagine it could be tricky.  Can we leave this for a future update?

tzanio commented on examples/ex1.cpp:
```diff
@@ -9,6 +9,8 @@
 //               ex1 -m ../data/fichera.mesh
 //               ex1 -m ../data/fichera-mixed.mesh
 //               ex1 -m ../data/toroid-wedge.mesh
+//               ex1 -m ../data/annulus-pi-3.msh
```
These need to be updated too

mlstowell replied:
Oops...

mlstowell replied:
I fixed the names in `ex1`, `ex1p`, and `ex11p`.  Also, check my spelling by comps-and-pasting the sample command lines.  They all seem to work as written.

tzanio commented on CHANGELOG:
```diff
@@ -29,6 +29,10 @@ Meshing improvements
 
 - Added TMOP capability for approximate tangential mesh relaxation.
 
+- Added support for reading periodic meshes written in Gmsh's version 2.2 file
+  format.  Also, including two Gmsh input files and corresponding ".msh" output
+  files for validating this new functionality.
```
```suggestion
- Added support for reading periodic meshes in Gmsh format (version 2.2). See
  for example the periodic-annulus-sector and periodic-torus-sector files in
  the data directory.
```

tzanio commented:
> I'm not sure about curved periodic meshes. I don't think I've seen Gmsh's curved mesh format. I imagine it could be tricky. Can we leave this for a future update?

They do support high-order elements, but I agree we should leave this for a different PR.
